### PR TITLE
abciClient.BeginBlockSync should not hang on crashed server 

### PR DESCRIPTION
**Pull request by ethanfrey**
* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG.md

Addressing issue #1890

Reproduced the errant behavior, and pushing to demo that.
Working on a solution.

### CODE REVIEW

File: abci/client/socket_client_test.go [hunk 1]
```diff
@@ -26,3 +33,94 @@ func TestSocketClientStopForErrorDeadlock(t *testing.T) {
...
+		require.True(t, ok, "Must not close channel")
+		assert.Error(t, err, "We should get EOF error")
+	}
+}
+
+func setupClientServer(t *testing.T, app types.Application) (
+	cmn.Service, abcicli.Client) {
+	// some port between 20k and 30k
+	port := 20000 + cmn.RandInt32()%10000
+	addr := fmt.Sprintf("localhost:%d", port)
```
Comment (melekes) [inline abci/client/socket_client_test.go]: can't we just use `:0` as port here, so server would pick ephemeral port for us?

Comment (melekes) [reply abci/client/socket_client_test.go]: I think the server should expose `Addr()` method, but this is for another PR. thanks

```
func (s *Server) Addr() net.Addr
Addr returns the server's listener's network address, either a *TCPAddr or *UnixAddr.
```

File: abci/client/socket_client_test.go [hunk 2]
```diff
@@ -26,3 +33,94 @@ func TestSocketClientStopForErrorDeadlock(t *testing.T) {
...
+	defer c.Stop()
+
+	resp := make(chan error, 1)
+	go func() {
+		// Start BeginBlock and flush it
+		reqres := c.BeginBlockAsync(types.RequestBeginBlock{})
+		flush := c.FlushAsync()
+		// wait 20 ms for all events to travel socket, but
+		// no response yet from server
+		time.Sleep(20 * time.Millisecond)
```
Comment (ebuchman) [inline abci/client/socket_client_test.go]: this seems non-deterministic. is there some way we can test this without depending on sleeps ?

File: abci/client/socket_client_test.go [hunk 3]
```diff
@@ -26,3 +33,94 @@ func TestSocketClientStopForErrorDeadlock(t *testing.T) {
...
+		// kill the server, so the connections break
+		s.Stop()
+
+		// wait for the response from BeginBlock
+		fmt.Println("waiting for begin block")
```
Comment (ebuchman) [inline abci/client/socket_client_test.go]: can we get rid of the print statements please?

File: abci/client/socket_client_test.go (hunk 2 above)
Comment (ethanfrey) [reply abci/client/socket_client_test.go]: I couldn't think of any way without major changes, as we wait for another goroutine to process an event and there is no clean way to hook into some event system.

Good thing to keep in mine for a refactor of this code.

File: abci/client/socket_client_test.go (hunk 3 above)
Comment (ethanfrey) [reply abci/client/socket_client_test.go]: Yeah, my bad.

I'm away from coding machines for a few days. Can you just push that onto this branch? I think I gave maintainers push access.

Comment (melekes) [reply abci/client/socket_client_test.go]: Done in https://github.com/tendermint/tendermint/pull/1950

**Pull request merged**